tests/components/tasmota/test_device_trigger.py

```diff
@@ -0,0 +1,832 @@
+    await hass.async_block_till_done()
+
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+
+
+async def test_attach_remove(hass, device_reg, mqtt_mock, setup_tasmota):
+    """Test attach and removal of trigger."""
+    # Discover a device with device trigger
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["swc"][0] = 0
+    mac = config["mac"]
+
+    mqtt_mock.async_subscribe.reset_mock()
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config))
+    await hass.async_block_till_done()
+
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+
+    calls = []
+
+    def callback(trigger, context):
+        calls.append(trigger["trigger"]["description"])
+
+    remove = await async_attach_trigger(
+        hass,
+        {
+            "platform": "device",
+            "domain": DOMAIN,
+            "device_id": device_entry.id,
+            "discovery_id": "00000049A3BC_switch_1_TOGGLE",
+            "type": "button_short_press",
+            "subtype": "switch_1",
+        },
+        callback,
+        None,
+    )
+
+    # Fake short press.
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+    assert calls[0] == "event 'tasmota_event'"
+
+    # Remove the trigger
+    remove()
+    await hass.async_block_till_done()
+
+    # Verify the triggers are no longer active
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 1
+
+
+async def test_attach_remove_late(hass, device_reg, mqtt_mock, setup_tasmota):
+    """Test attach and removal of trigger."""
+    # Discover a device without device triggers
+    config1 = copy.deepcopy(DEFAULT_CONFIG)
+    config1["swc"][0] = -1
+    mac = config1["mac"]
+
+    # Discover a device with device triggers
+    config2 = copy.deepcopy(DEFAULT_CONFIG)
+    config2["swc"][0] = 0
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", json.dumps(config1))
+    await hass.async_block_till_done()
+
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+
+    calls = []
+
+    def callback(trigger, context):
+        calls.append(trigger["trigger"]["description"])
+
+    remove = await async_attach_trigger(
+        hass,
+        {
+            "platform": "device",
+            "domain": DOMAIN,
+            "device_id": device_entry.id,
+            "discovery_id": "00000049A3BC_switch_1_TOGGLE",
+            "type": "button_short_press",
+            "subtype": "switch_1",
+        },
+        callback,
+        None,
+    )
+
+    # Fake short press.
+    async_fire_mqtt_message(hass, "tasmota_49A3BC/stat/SWITCH1T", '{"TRIG":"TOGGLE"}')
+    await hass.async_block_till_done()
+    assert len(calls) == 0
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)